parser/transaction.go

```diff
@@ -397,76 +382,207 @@ func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
+	}
+	if !s.ReadCompactSize(&outputCount) {
+		return nil, errors.New("could not read nShieldedOutput")
+	}
+	tx.shieldedOutputs = make([]*output, outputCount)
+	for i := 0; i < outputCount; i++ {
+		newOutput := &output{}
+		s, err = newOutput.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Output")
+		}
+		tx.shieldedOutputs[i] = newOutput
+	}
+	if spendCount+outputCount > 0 && !s.ReadInt64(&tx.valueBalanceSapling) {
+		return nil, errors.New("could not read valueBalance")
+	}
+	if spendCount > 0 && !s.Skip(32) {
+		return nil, errors.New("could not skip anchorSapling")
+	}
+	if !s.Skip(192 * spendCount) {
+		return nil, errors.New("could not skip vSpendProofsSapling")
+	}
+	if !s.Skip(64 * spendCount) {
+		return nil, errors.New("could not skip vSpendAuthSigsSapling")
+	}
+	if !s.Skip(192 * outputCount) {
+		return nil, errors.New("could not skip vOutputProofsSapling")
+	}
+	if spendCount+outputCount > 0 && !s.ReadBytes(&tx.bindingSigSapling, 64) {
+		return nil, errors.New("could not read bindingSigSapling")
+	}
+	var actionsCount int
+	if !s.ReadCompactSize(&actionsCount) {
+		return nil, errors.New("could not read nActionsOrchard")
+	}
+	if !s.Skip(820 * actionsCount) {
+		return nil, errors.New("could not skip vActionsOrchard")
+	}
+	if actionsCount > 0 {
+		if !s.Skip(1) {
+			return nil, errors.New("could not skip flagsOrchard")
+		}
+		if !s.Skip(8) {
+			return nil, errors.New("could not skip valueBalanceOrchard")
+		}
+		if !s.Skip(32) {
+			return nil, errors.New("could not skip anchorOrchard")
+		}
+		var proofsCount int
+		if !s.ReadCompactSize(&proofsCount) {
+			return nil, errors.New("could not read sizeProofsOrchard")
+		}
+		if !s.Skip(proofsCount) {
+			return nil, errors.New("could not skip proofsOrchard")
+		}
+		if !s.Skip(64 * actionsCount) {
+			return nil, errors.New("could not skip vSpendAuthSigsOrchard")
+		}
+		if !s.Skip(64) {
+			return nil, errors.New("could not skip bindingSigOrchard")
 		}
 	}
+	return s, nil
+}
 
+// ParseFromSlice deserializes a single transaction from the given data.
+func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
+	s := bytestring.String(data)
+
+	// declare here to prevent shadowing problems in cryptobyte assignments
+	var err error
+
+	var header uint32
+	if !s.ReadUint32(&header) {
+		return nil, errors.New("could not read header")
+	}
+
+	tx.fOverwintered = (header >> 31) == 1
+	tx.version = header & 0x7FFFFFFF
+
+	// Implement the effective version rule. From the spec section 7.1:
+	//
+	// "Version constraints apply to the effectiveVersion, which is equal to
+	// min(2, version) when fOverwintered = 0 and to version otherwise."
+	if !tx.fOverwintered && tx.version > 2 {
+		tx.version = 2
+	}
+
+	// Spec says fOverwinter must be set for version 5.
+	if tx.version >= 5 && !tx.fOverwintered {
+		return nil, errors.New(fmt.Sprintf("version %d requires fOverwinter", tx.version))
+	}
```

Comment:
Thank you for this list, I missed this part of the spec (section 7.1.2). I just realized that I can make the checks even tighter because lightwalletd only works with Sapling and beyond. So, for example, the version number must be 4 or greater. The parser can error on seeing a lower version number. (When lightwalletd starts up for the first time, it queries the Sapling activation height from `zcashd` and starts download blocks from that height.)

Reply:
There are a bunch more rules in 7.1.2, such as
> `nSpendsSapling`, `nOutputsSapling`, and `nActionsOrchard` MUST all be less than 2^16

I don't think we have to check all these rules, because we already trust our connected `zcashd` in many ways. But I think it's good to check these initial transaction fields (`fOverwintered`, `version`, `nVersionGroupID`) because that will catch situations where we're not actually talking to a `zcashd`, or we're on the wrong network, or something is else is seriously wrong, and we're getting random garbage. But beyond these simple, obvious sanity checks, I don't think we have to check everything (we haven't in the past). We could check more if and when we move to a more hardened security model. Does this seem reasonable?